File: lua/test/deadline_entity_test.lua
-- Deadline entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("dedline_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("DeadlineEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Deadline(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = deadline_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "deadline." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set DEDLINE_TEST_DEADLINE_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local deadline_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.deadline")))
    local deadline_ref01_data = nil
    if #deadline_ref01_data_raw > 0 then
      deadline_ref01_data = helpers.to_map(deadline_ref01_data_raw[1][2])
    end

    -- LIST
    local deadline_ref01_ent = client:Deadline(nil)
    local deadline_ref01_match = {}

    local deadline_ref01_list_result, err = deadline_ref01_ent:list(deadline_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(deadline_ref01_list_result)

  end)
end)

function deadline_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/deadline/DeadlineTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read deadline test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "deadline01", "deadline02", "deadline03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("DEDLINE_TEST_DEADLINE_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["DEDLINE_TEST_DEADLINE_ENTID"] = idmap,
    ["DEDLINE_TEST_LIVE"] = "FALSE",
    ["DEDLINE_TEST_EXPLAIN"] = "FALSE",
    ["DEDLINE_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["DEDLINE_TEST_DEADLINE_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["DEDLINE_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["DEDLINE_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["DEDLINE_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["DEDLINE_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: php/test/DeadlineEntityTest.php
<?php
declare(strict_types=1);

// Deadline entity test

require_once __DIR__ . '/../dedline_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class DeadlineEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = DedlineSDK::test(null, null);
        $ent = $testsdk->Deadline(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "deadline" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = DedlineSDK::test($seed, null);
        $seen = iterator_to_array($base->Deadline(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = DedlineConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = DedlineSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Deadline(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = deadline_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "deadline." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set DEDLINE_TEST_DEADLINE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $deadline_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.deadline")));
        $deadline_ref01_data = null;
        if (count($deadline_ref01_data_raw) > 0) {
            $deadline_ref01_data = Helpers::to_map($deadline_ref01_data_raw[0][1]);
        }

        // LIST
        $deadline_ref01_ent = $client->Deadline(null);
        $deadline_ref01_match = [];

        $deadline_ref01_list_result = $deadline_ref01_ent->list($deadline_ref01_match, null);
        $this->assertIsArray($deadline_ref01_list_result);

    }
}

function deadline_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/deadline/DeadlineTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = DedlineSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["deadline01", "deadline02", "deadline03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("DEDLINE_TEST_DEADLINE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "DEDLINE_TEST_DEADLINE_ENTID" => $idmap,
        "DEDLINE_TEST_LIVE" => "FALSE",
        "DEDLINE_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["DEDLINE_TEST_DEADLINE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["DEDLINE_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new DedlineSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["DEDLINE_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["DEDLINE_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
